### ISSUE TITLE
Code execution commands

### ISSUE THREAD

**Issue by davidanthoff** [julia-vscode/julia-vscode#1221]
This is an attempt to come up with a more or less comprehensive list of commands we should have for code execution. I'm leaving out key bindings for now, I think that is a complicated topic we can tackle later, for now it would just be nice if we had a complete set of commands.

I think we have three different dimensions in this problem, and then we should offer one command for all possible combinations, taking out combinations that don't make sense (e.g. run whole file and move cursor).

1. WHAT to run: a) selected text, b) code block, c) code cell, d) everything above cursor position, e) whole file, e) auto mode (if text is selected => run that, if cells is defined => run cell, else => run code block)
2. CURSOR movement: a) don't move cursor, b) move cursor to next cell/block
3. DEBUG mode: a) no debugging, b) enter debugger, c) run debugger

Or something like that. Maybe we could first try to complete this list, i.e. add more items that come to mind. Then we could write down all permutations and check which ones make sense and whether we missed some aspect, and then we could start to implement this?

**Comment by pfitzseb** [comment]
In Juno we currently have (also added in the current keybindings):
- Run Block (either runs the currently selected text or the block around the cursor) <kbd>Ctrl-Enter</kbd>
- Run Block and Move <kbd>Shift-Enter</kbd>
- Run Cell <kbd>Alt-Enter</kbd>
- Run Cell and Move <kbd>Alt-Shift-Enter</kbd>
- Run File <kbd>Ctrl-Shift-Enter</kbd>
- Debug Block
- Step Through Block
- Debug Cell
- Step Through Cell
- Debug File
- Step Through File

That's pretty close to what you listed above, but
- has only one ("auto mode") command for running the selected text or the current block
- doesn't have "run everything above cursor position"

IMHO having one command for selection/block/cell execution doesn't make too much sense because block- and cell-level execution have fairly distinct use cases.

One other somewhat related set of commands is for moving through blocks/cells:
- Next Cell <kbd>Alt-Down</kbd>
- Previous Cell <kbd>Alt-Up</kbd>
- Next Block (not in Juno)
- Previous Block (not in Juno)

**Comment by pfitzseb** [comment]
`Ctrl-Shift-J` is a possible keybinding for showing the "Julia Explorer" pane (which currently only has the workspace).

**Comment by pfitzseb** [comment]
Since we wanted to re-use the notebook keybindings here:
[The python notebook](https://code.visualstudio.com/docs/python/jupyter-support#_run-a-single-code-cell) use `Ctrl-Enter` for running the current cell and `Shift-Enter` for running the current cell and moving to the next one. This is the same as what I suggested above if you substitute "cell" with "block" (which makes sense, imho). The (`Shift-`)`Alt-Enter` keybindings for text editor cells don't really conflict either.

As for other keybindings (select module etc), we can hide those behind `Ctrl-I`, which seems to be the only unused `Ctrl+*letter*` keybinding in VSCode.

If there's no disagreement on this I'd like to open a PR for this fairly soon (so we can get it into 1.0).

**Comment by aviatesk** [comment]
As for the notebook keybindings:
I totally agree with @pfitzseb ideas. 1.) using <kbd>ctrl</kbd> and <kbd>alt</kbd> for block/cell based execution separately and respectively, and 2.) automatically respects selection if exists, seem to me like a "standard" in this kind of "in-editor canvas programming" -- e.g. [hydrogen](https://github.com/nteract/hydrogen), which offers a matured canvas programming environment powered by Jupyter, also has almost identical keybindings/logic (see [this file](https://github.com/nteract/hydrogen/blob/master/keymaps/hydrogen.cson) if interested).

For other "our extension specific" keybindings, can't we use <kbd>alt-j</kbd>-prefixed keys ? It's not used by vscode by default and having the <kbd>j</kbd> prefix sounds like more natural for those Julia related commands.

**Comment by davidanthoff** [comment]
I don't think the Python extension is what we want to look at, the functionality you are linking to is still their old, homegrown UI that is not based on the notebook UI that ships via VS Code itself.

I looked through the core VS Code repo and didn't find any discussion of keyboard shortcuts for the new notebook UI, so I opened a new issue asking for a status update: https://github.com/microsoft/vscode/issues/101436. They normally respond very quickly, so hopefully we'll have a response within a few days.

**Comment by davidanthoff** [comment]
I did hear back from MS: they are going to come up with final keyboard shortcuts for the notebook experience, but not in time for our Juliacon release/1.0 release. So I think we should defer any further changes here and stick with what we have, and then redo the keyboard shortcuts once the notebook keyboard story on MS' side is settled.

**Comment by NightMachinary** [comment]
`language-julia.selectBlock` currently selects rather big blocks. Can we have a select command that selects the smallest scope it's in? For example, in:
```
function big()
 function small()
  # caret here
 end
end
```
It'd select the `small` function. In:
```
function big()
 function small()
  let a=2
   # caret here
  end
 end
end
```
It'd select the `let` block.

This will help with using the REPL effectively.

**Comment by jebej** [comment]
Is it possible to configure the keybindings? Right now the execute code block in file context is bound to `alt-enter`, instead of `ctrl-enter`, which I would prefer (like in Atom).

**Comment by pfitzseb** [comment]
Yes: https://github.com/julia-vscode/julia-vscode/wiki/Keyboard-shortcuts#replicating-juno-shortcuts.

**Comment by jebej** [comment]
thanks!
